File: main/src/main/java/com/sedmelluq/discord/lavaplayer/format/transcoder/OpusChunkEncoder.java
package com.sedmelluq.discord.lavaplayer.format.transcoder;

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat;
import com.sedmelluq.discord.lavaplayer.natives.opus.OpusEncoder;
import com.sedmelluq.discord.lavaplayer.player.AudioConfiguration;

import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

/**
 * Audio chunk encoder for Opus codec.
 */
public class OpusChunkEncoder implements AudioChunkEncoder {
  private final AudioDataFormat format;
  private final OpusEncoder encoder;
  private final ByteBuffer encodedBuffer;

  /**
   * @param configuration Audio configuration used for configuring the encoder
   * @param format Target audio format.
   */
  public OpusChunkEncoder(AudioConfiguration configuration, AudioDataFormat format) {
    encodedBuffer = ByteBuffer.allocateDirect(format.maximumChunkSize());
    encoder = new OpusEncoder(format.sampleRate, format.channelCount, configuration.getOpusEncodingQuality());
    this.format = format;
  }

  @Override
  public byte[] encode(ShortBuffer buffer) {
    encoder.encode(buffer, format.chunkSampleCount, encodedBuffer);

    byte[] bytes = new byte[encodedBuffer.remaining()];
    encodedBuffer.get(bytes);
    return bytes;
  }

  @Override
  public void encode(ShortBuffer buffer, ByteBuffer outBuffer) {
    if (outBuffer.isDirect()) {
      encoder.encode(buffer, format.chunkSampleCount, outBuffer);
    } else {
      encoder.encode(buffer, format.chunkSampleCount, encodedBuffer);

      int length = encodedBuffer.remaining();
      encodedBuffer.get(outBuffer.array(), 0, length);

      outBuffer.position(0);
      outBuffer.limit(length);
    }
  }

  @Override
  public void close() {
    encoder.close();
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/format/transcoder/OpusChunkEncoder.kt
package com.sedmelluq.discord.lavaplayer.format.transcoder

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat
import com.sedmelluq.discord.lavaplayer.manager.AudioConfiguration
import com.sedmelluq.discord.lavaplayer.natives.opus.OpusEncoder
import java.nio.ByteBuffer
import java.nio.ShortBuffer

/**
 * Audio chunk encoder for Opus codec.
 *
 * @param configuration Audio configuration used for configuring the encoder
 * @param format        Target audio format.
 */
class OpusChunkEncoder(configuration: AudioConfiguration, private val format: AudioDataFormat) : AudioChunkEncoder {
    private val encodedBuffer = ByteBuffer.allocateDirect(format.maximumChunkSize)
    private val encoder = OpusEncoder(format.sampleRate, format.channelCount, configuration.opusEncodingQuality)

    override fun encode(buffer: ShortBuffer): ByteArray {
        encoder.encode(buffer, format.chunkSampleCount, encodedBuffer)
        val bytes = ByteArray(encodedBuffer.remaining())
        encodedBuffer.get(bytes)
        return bytes
    }

    override fun encode(input: ShortBuffer, output: ByteBuffer) {
        if (output.isDirect) {
            encoder.encode(input, format.chunkSampleCount, output)
        } else {
            encoder.encode(input, format.chunkSampleCount, encodedBuffer)
            val length = encodedBuffer.remaining()
            encodedBuffer[output.array(), 0, length]
            output.position(0)
            output.limit(length)
        }
    }

    override fun close() {
        encoder.close()
    }
}
